### PR TITLE
feat(dashboard): extend breakdown workspace and refresh UX

### PR DESCRIPTION
**Pull request by SeaXen**
## Summary
This PR is the consolidated dashboard update for `hermes-telemetry`.

It includes the earlier Breakdown workspace work from #25, the refresh/session-visibility follow-up, audit hardening fixes, and a new operator-forensics pass that splits dense forensic surfaces into dedicated Request / Tool / Error pages.

### 1) Breakdown workspace + investigation UX
- add the full telemetry **Breakdown** workspace introduced in #25
- split the UI into **Home** and **Breakdown** views
- add model token usage, daily token usage, cost columns, and the combined **Daily Token + Model Breakdown** visualization
- add **Provider Cost + Token Breakdown** and **Cache Efficiency Stats** sections
- add **Session Investigation** with:
  - provider/model/day/status/platform/cron/tool drill-down
  - provider-chart click-through
  - removable filter chips
  - search box
  - session detail drawer with clickable chips back into investigation filters
- remove the **Cron Job** column from Investigation Workspace for a cleaner filtered session table

### 2) Operator forensics pages
- add dedicated internal tabs: **Request**, **Tool**, and **Error**
- move request-level telemetry out of Breakdown into the **Request** page
- move tool usage/failure analytics out of Breakdown into the **Tool** page
- move grouped incident/error views out of Breakdown into the **Error** page
- add request detail drawer with request metrics, sibling LLM calls, session totals, and session tool activity
- add **Provider Health / Anomaly Signals**:
  - current calls vs previous equal window
  - traffic delta
  - estimated-call percentage
  - failure percentage
  - latency delta
  - anomaly labels such as traffic spike, latency spike, estimated-heavy, and run failures
- tighten provider/model rendering into single-line chips across dense tables and drawers
- remove the **Reasoning** column from **LLM Request Forensics** to reduce table width
- remove the **Cron Job** column from **Recent Sessions** to reduce noise on Home

### 3) Refresh/loading UX
- add header **auto-refresh** controls: `Off / 5s / 10s / 20s / 1min`
- add header load status text
- improve first-load shell and non-blanking background refresh behavior
- add stale async response guards for both:
  - `loadAll()`
  - `openSessionDetail()`

### 4) Session visibility behavior
- soft-hide deleted Hermes sessions by default in dashboard session lists
- preserve telemetry totals/history
- fail open safely if Hermes session metadata is unavailable, so non-cron telemetry is not accidentally hidden

### 5) Audit-driven fixes in this PR
A final live audit found and fixed several real issues:

- **Correctness:** fix bounded time-window filtering so dashboard APIs compare timestamps using the same UTC ISO format stored in `telemetry.db`
- **Correctness:** fix daily cost chart window filtering to use the same timestamp semantics
- **Budget UX:** fix monthly budget editing so the dashboard uses `global/monthly` consistently and the modal preloads the correct window
- **Budget semantics:** align dashboard budget window calculations with actual budget enforcement windows (current local day/month start rather than rolling 24h/30d)
- **Robustness:** return structured `400` JSON responses for malformed numeric query params instead of dropping the connection
- **Security:** remove wildcard `Access-Control-Allow-Origin: *` from telemetry JSON responses to avoid cross-origin browser reads of local telemetry
- **Frontend safety:** escape remaining dynamic HTML fields in dashboard tables/badges

## Why this PR replaces #25
This branch contains all of the #25 work and then extends it with the newer UX, hardening, and operator-forensics fixes.

So reviewers only need to review **#26**.

PR #25 was closed as superseded to avoid duplicate review on an older subset.

### Automated checks
- `python3 -m py_compile dashboard/serve.py`
- `uv run --with ruff --no-project ruff check dashboard/serve.py tests/test_dashboard.py tests/test_stats_models.py tests/test_stats_providers.py`
- `uv run --with ruff --no-project ruff format --check dashboard/serve.py tests/test_dashboard.py tests/test_stats_models.py tests/test_stats_providers.py`
- `PYTHONPATH=. uv run --with pytest --with watchdog --with pyyaml --no-project python -m pytest -q tests/test_dashboard.py tests/test_stats_models.py tests/test_stats_providers.py`

### Result
- **35 passed**
- ruff check passed
- ruff format-check passed
- `git diff --check` passed
- secret-pattern scan of staged diff found no credentials

### Live runtime verification
- `systemctl restart hermes-telemetry-dashboard`
- verified dashboard service active on `0.0.0.0:8009`
- verified in browser at `http://127.0.0.1:8009`
- browser console: **0 JS errors**

### UI validation
- Home **Recent Sessions** table no longer has `Cron Job`
- Request **LLM Request Forensics** table no longer has `Reasoning`
- Request page shows:
  - `Provider Health / Anomaly Signals`
  - `LLM Request Forensics`
- Tool page shows:
  - `Tool Analytics`
- Error page shows:
  - `Error Center`
  - `Run Status Groups`
  - `Failed Tools`
  - `Recent Incidents`
- Breakdown no longer duplicates Request / Tool / Error sections
- provider/model chips stay single-line and readable in recent, investigation, request, and drawer views

### Direct validation of audit fixes
- `/api/summary?hours=24` now matches a direct SQLite query for the same UTC ISO cutoff window
- `/api/budget` reports:
  - `global/daily`
  - `global/monthly`
- monthly budget modal opens with:
  - `scope=global`
  - `window=monthly`
  - correct prefilled monthly limit
- malformed query example `/api/summary?hours=abc` returns:
  - HTTP `400`
  - JSON body: `{"error": "invalid hours: 'abc'"}`
- telemetry JSON responses no longer send `Access-Control-Allow-Origin: *`

## Known limitation
- full test suite still has the **pre-existing upstream** failure in `tests/test_db.py::test_schema_idempotent`
- targeted dashboard-related checks pass
- that full-suite failure is outside the scope of this PR

## Notes for reviewer
- #26 is the full replacement for #25
- review this PR only; #25 is a strict subset
- latest commits on this branch include:
  - `ca4a8be` — `feat(dashboard): add operator forensics pages`
  - `140ae9d` — `fix(dashboard): harden telemetry windows and budget editor`

## Latest follow-up batch

Added and verified:
- **Model Efficiency Score** on **Breakdown**
  - provider/model scorecard using cost per 1M tokens, output/input ratio, cache hit share, tool calls per API call, failure %, and latency
- **Tool Failure Heatmap** on **Tool**
  - grouped by tool × model × platform × cron with fail %, failed calls, latency, and last seen
- **Cron Failure / Waste Center** on **Error**
  - cron-focused failure/waste view with fail %, timeout count, failed tool calls, wasted tokens/cost, last-success age, and avg duration
- Removed duplicate in-content page-title header above page sections; header tabs remain the single page navigation label

## Hardening in this batch

- Fixed inline handler escaping so drilldown/session/request actions safely escape provider/model/tool/status/platform/session/cron values used inside `onclick` attributes
- Tightened aggregate subqueries so new failure/efficiency metrics respect the selected time window and visible-session filtering
- Clamped new API `limit` params to sane bounds

## Final verification

- `python3 -m py_compile dashboard/serve.py` ✅
- `PYTHONPATH=. uv run --with pytest --with watchdog --with pyyaml --no-project python -m pytest -q tests/test_dashboard.py tests/test_stats_models.py tests/test_stats_providers.py` → **36 passed** ✅
- `uv run --with ruff --no-project ruff check dashboard/serve.py tests/test_dashboard.py tests/test_stats_models.py tests/test_stats_providers.py` ✅
- `uv run --with ruff --no-project ruff format --check dashboard/serve.py tests/test_dashboard.py tests/test_stats_models.py tests/test_stats_providers.py` ✅
- Independent review after fixes: **passed** ✅
- Live dashboard restart + browser verification on `http://127.0.0.1:8009` ✅
  - Breakdown shows **Model Efficiency Score**
  - Tool shows **Tool Failure Heatmap**
  - Error shows **Cron Failure / Waste Center**
  - Duplicate page-title header removed
  - Browser console: **0 JS errors**

## Model usage trend batch

Added and verified on **Breakdown**:
- **Model Usage Trends** card
  - grouping toggle: `Daily / Weekly / Monthly`
  - metric toggle: `Tokens / Cost / Requests / Share %`
  - stacked per-model chart
- **Model Share + Delta** card
  - current vs previous period comparison
  - per-model token share, previous share, share delta, cost share, and request count

## Fixes and hardening in this batch

- prevent fake comparison rows when only one period exists; show a proper empty state instead
- fix weekly grouping to bucket by **week-start Monday date** so year-boundary weeks are grouped correctly
- limit daily drilldown clicks to the day view only; week/month trend views no longer imply unsupported period drilldown
- add explicit empty-state messaging for sparse datasets
- keep earlier live-verified table cleanup intact:
  - Home **Recent Sessions** has no `Cron Job` column
  - Request **LLM Request Forensics** has no `Reasoning` column

## Final verification for this batch

- `python3 -m py_compile dashboard/serve.py` ✅
- `uv run --no-project --with pytest python -m pytest -q tests/test_dashboard.py` → **17 passed** ✅
- `uv run --no-project --with ruff ruff check dashboard/serve.py tests/test_dashboard.py` ✅
- `git diff --check` ✅
- Independent review after weekly-bucketing fix: **passed** ✅
- `systemctl restart hermes-telemetry-dashboard` → service **active** ✅
- Live browser verification on `http://127.0.0.1:8009` ✅
  - Breakdown shows **Model Usage Trends** and **Model Share + Delta**
  - Daily / Weekly / Monthly and Tokens / Cost / Requests / Share % controls render and switch correctly
  - Weekly view shows safe empty-state comparison messaging when no previous week exists
  - Browser console: **0 JS errors**

### CODE REVIEW

File: dashboard/serve.py [hunk 1]
```diff
@@ -235,28 +1569,16 @@ def api_budget():
...
-    for win_key, win_label, since in [
-        ("daily", "global/daily", now - timedelta(hours=24)),
-        ("monthly", "monthly", now - timedelta(days=30)),
-    ]:
+    for win_key in ("daily", "monthly"):
         limit = g.get(f"{win_key}_usd")
         if limit is None:
             continue
         spend = _one(
             "SELECT COALESCE(SUM(cost_usd),0.0) AS spent, COALESCE(SUM(estimated_llm_calls),0) AS est, COALESCE(SUM(api_calls),0) AS total FROM runs WHERE started_at >= ?",
```
Comment (nujovich) [inline dashboard/serve.py:1814]: [minor] `_budget_window_start_utc` uses `datetime.now().astimezone()`, which reads the server's local timezone. When the dashboard is accessed from a different timezone, budget windows render shifted from what the viewer expects. Please make the displayed windows respect the viewer's timezone.

File: dashboard/serve.py [hunk 2]
```diff
@@ -502,7 +1994,6 @@ def _json(self, data, status=200):
         body = json.dumps(data, default=str).encode()
         self.send_response(status)
```
Comment (nujovich) [inline dashboard/serve.py:2259]: [minor] CORS `Access-Control-Allow-Origin: *` removed — genuine security improvement. Consider adding `Vary: Origin` if this dashboard might ever sit behind a cache.

**Review by nujovich — COMMENTED** [review]
## Automated review (harness)

Large, well-structured dashboard overhaul. The security fix (CORS `*` removal), input validation (`_parse_int` → 400), and budget window alignment with `_budget_window_start_utc` are solid. No blockers — ruff-clean and tests pass.

**Please address before merge:**

1. **README — document the new dashboard.** Update the README to cover the new tabbed layout (Home/Breakdown/Request/Tool/Error), auto-refresh, and the investigation workspace, and replace the outdated single-page screenshots with a current one.

2. **Timezone-aware dates.** When the dashboard is opened from a machine in a different timezone than the server, budget windows currently render in the server's local time. Please make the displayed dates/times respect the viewer's timezone instead of the server's.

3. **Disclose hidden sessions.** The new behavior that hides sessions marked as deleted in Hermes metadata should be documented in the README, so users understand why some sessions don't appear instead of seeing unexplained gaps.

**Review by nujovich — CHANGES_REQUESTED** [review]

**Comment by SeaXen** [comment]
Addressed the maintainer feedback on this branch.

Fixed in this push:
- updated `README.md` for the current tabbed dashboard layout (`Home / Breakdown / Request / Tool / Error`)
- documented header auto-refresh behavior and the investigation workspace
- documented the soft-hidden deleted-session behavior in session-facing tables
- replaced the outdated dashboard screenshot with a current dashboard overview image
- made dashboard-rendered timestamps viewer-local in the browser
- made budget windows timezone-aware for the viewer (`tz` passed from the dashboard to `/api/budget` and `/api/budget/detail`)
- preserved viewer-timezone metadata on `POST /api/budget` responses too
- restored default threshold values in `api_budget_detail()` for partial configs so the budget modal keeps its old safe defaults

Verification:
- `python3 -m py_compile dashboard/serve.py`
- `uv run --no-project --with pytest python -m pytest -q tests/test_dashboard.py` → 21 passed
- `uv run --no-project --with ruff ruff check dashboard/serve.py tests/test_dashboard.py`
- live browser verification on `http://127.0.0.1:8009`
- browser POST check: `/api/budget?tz=UTC` returns UTC window metadata
- independent review after fixes: passed

**Comment by SeaXen** [comment]
Final recheck complete on head `a80cad4`.

Addressed from the maintainer review:
- README updated for the current tabbed dashboard layout, auto-refresh behavior, investigation workspace, and hidden-session behavior
- dashboard screenshot refreshed
- viewer-local timestamp rendering added in the UI
- budget windows now use the viewer timezone
- `POST /api/budget` now also preserves viewer-timezone metadata
- `api_budget_detail()` restored default threshold values for partial configs

Re-verified after the follow-up fixes:
- `python3 -m py_compile dashboard/serve.py`
- `uv run --no-project --with pytest python -m pytest -q tests/test_dashboard.py` → 21 passed
- `uv run --no-project --with ruff ruff check dashboard/serve.py tests/test_dashboard.py`
- live browser check on `http://127.0.0.1:8009`
- direct browser POST check for `/api/budget?tz=UTC` returned UTC window metadata
- independent final review pass: clean

From my side this batch is ready. The only thing still not updated is GitHub’s review state, which remains `CHANGES_REQUESTED` until maintainer re-reviews.

**Review by nujovich — APPROVED** [review]
## Automated review (harness)

Comprehensive and well-tested dashboard overhaul. The changes are scoped entirely to `dashboard/` and its tests — no core plugin invariants (pricing engine, budget enforcement, hook pipeline) are touched.

**What's good:**
- Tabbed navigation (Home / Breakdown / Request / Tool / Error) with operator-forensics pages is a clean information architecture
- Security hardening: CORS wildcard removed, HTML escaping across dynamic fields, `jsAttr()` for onclick safety, `_parse_int` for input validation returning structured 400s
- Viewer-local timezone budget windows with proper UTC window bounds
- Async stale-response guards (`latestLoadRequestId`, `latestDrawerRequestId`)
- Soft-deleted session filtering with safe fallback when Hermes metadata is unavailable
- Auto-refresh controls saved to localStorage
- Tests: 293 new lines in `test_dashboard.py` covering new API endpoints, budget window math, model trends, error center, and edge cases (single-period comparison, week-boundary grouping)

**Minor notes:**
- `/api/budget` scope key changed from plain window label to `global/{win_key}` — frontend is updated in this PR, but any hypothetical external dashboard consumer would need the same update
- The pre-existing `test_schema_idempotent` failure is noted as outside scope

Approving — this is ready to merge.
